pil-analyzer/src/type_inference.rs

```diff
@@ -441,31 +448,29 @@ impl<'a> TypeChecker<'a> {
         expected_type: &ExpectedType,
         expr: &mut Expression,
     ) -> Result<(), Error> {
-        if expected_type.allow_array {
-            self.infer_type_of_expression(expr).and_then(|ty| {
-                let ty = self.type_into_substituted(ty);
-                let expected_type = if matches!(ty, Type::Array(_)) {
-                    Type::Array(ArrayType {
-                        base: Box::new(expected_type.ty.clone()),
-                        length: None,
-                    })
-                } else {
-                    expected_type.ty.clone()
-                };
+        self.infer_type_of_expression(expr).and_then(|ty| {
```

Comment:
I think this could now also just be `let ty = self.infer_type_of_expression(expr)?;`